cibyl/sources/source.py

```diff
@@ -50,38 +53,70 @@ def request_handler(*args, **kwargs):
     return request_handler
 
 
-class Source:
-    """Represents a source of a system on which queries are performed."""
+class Source(AttrDict):
+    """Represents a data provider within a system."""
 
-    def __init__(self, name: str, driver: str, url: str = None,
-                 priority: int = 0):
-        self.name = name
-        self.driver = driver
-        self.url = url
-        self.priority = priority
+    def __init__(self, name: str, driver: str, **kwargs):
+        kwargs.setdefault('enabled', True)
+        kwargs.setdefault('priority', 0)
 
-    @staticmethod
-    def get_source_method(system_name: str, sources: list, func_name: str):
-        """Returns a method of a single source given all the sources
-        of the system and the name of function.
+        super().__init__(name=name, driver=driver, **kwargs)
 
-        An exception is raised if there are no sources with such function
-        name or if there are multiple sources that have this function.
 
-        :param system_name: The name of system
-        :type system_name: str
-        :param sources: List of Source instances
-        :type sources: list[Source]
-        :param func_name: The name of the function to invoke
-        :type func_name: str
-        """
-        valid_sources = []
+def is_source_valid(source: Source, desired_attr: str):
+    """Checks if a source can be considered valid to perform a query.
+
+    For a source to be considered valid it must:
+        * Be enabled.
+        * Have the attribute passed as input.
+
+    :param source: The source to check.
+    :type source: :class:`Source`
+    :param desired_attr: An attribute that is useful for performing a
+        query and that is desired for the source to have.
+    :type desired_attr: str
+    :return: Whether the source is valid or not.
+    :rtype: bool
+    """
+    if not source.enabled:
+        return False
+
+    if not hasattr(source, desired_attr):
+        return False
+
+    return True
+
+
+def get_source_method(system_name: str, sources: list, func_name: str):
+    """Returns a method of a single source given all the sources
+    of the system and the name of function.
+
+    An exception is raised if there are no sources with such function
+    name or if there are multiple sources that have this function.
+
+    :param system_name: The name of system
+    :type system_name: str
+    :param sources: List of Source instances
+    :type sources: list[Source]
+    :param func_name: The name of the function to invoke
+    :type func_name: str
+    """
+
+    def get_valid_sources():
```

Comment:
Is there any advantage to define this function rather than just renaming result to valid_sources?

Reply:
Not in particular, it is just my way of organizing the code. If we find it uncomfortable I can do it the other way around.